### PR TITLE
Feature/const global cache

### PR DESCRIPTION
**Pull request by kidder**

### CODE REVIEW

File: src/Parallel/ConstGlobalCache.hpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+namespace Parallel {
+
+/// \ingroup Parallel
+/// A Charm++ chare that caches constant data once per Charm++ node.
+///
+/// Metavariables must define the following metavariables:
+///   - const_global_cache_tag_list   typelist of tags of constant data
+///   - tentacle_list   typelist of Tentacles
+///
+/// \tparam Metavariables a struct containing metavariables
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: remove

File: src/Parallel/ConstGlobalCache.hpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+                         value> = nullptr>
+  typename TentacleTag::type& get() {
+    return tuples::get<TentacleTag>(tentacles_);
+  }
+
+  template <typename TentacleTag,
+            Requires<tmpl::found<
+                tentacle_list, std::is_same<tmpl::_1, tmpl::pin<TentacleTag>>>::
+                         value> = nullptr>
+  const typename TentacleTag::type& get() const {
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: get_tentacle (remove Requires)

File: src/Parallel/ConstGlobalCache.hpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  ///
+  /// \requires TentacleTag is a tag in tentacle_list
+  ///
+  /// \returns a Charm++ proxy that can be used to call an entry method on the
+  /// chare(s)
+  template <typename TentacleTag,
+            Requires<tmpl::found<
+                tentacle_list, std::is_same<tmpl::_1, tmpl::pin<TentacleTag>>>::
+                         value> = nullptr>
+  typename TentacleTag::type& get() {
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: get_tentacles (and remove Requires)

File: src/Parallel/ConstGlobalCache.hpp [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+          tt::is_a_v<std::unique_ptr, typename ConstGlobalCacheTag::type>> =
+          nullptr>
+  const typename ConstGlobalCacheTag::type::element_type& get() const {
+    return *(tuples::get<ConstGlobalCacheTag>(const_global_cache_).get());
+  }
+
+  template <
+      typename ConstGlobalCacheTag,
+      Requires<
+          tmpl::found<
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: remove found

File: src/Parallel/ConstGlobalCache.hpp [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+  /// \brief Access data in the cache
+  ///
+  /// \requires ConstGlobalCacheTag is a tag in tag_list
+  ///
+  /// \returns a constant reference to an object in the cache
+  template <
+      typename ConstGlobalCacheTag,
+      Requires<
+          tmpl::found<
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: remove found

File: src/Parallel/ConstGlobalCache.hpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            Requires<tmpl::found<
+                tentacle_list, std::is_same<tmpl::_1, tmpl::pin<TentacleTag>>>::
+                         value> = nullptr>
+  const typename TentacleTag::type& get() const {
+    return tuples::get<TentacleTag>(tentacles_);
+  }
+
+  /// Entry method to set the Tentacles (should only be called once)
+  void set_tentacles(tuples::TaggedTupleTypelist<tentacle_list>& tentacles,
+                     CkCallback callback);
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: Callback&

File: src/Parallel/ConstGlobalCache.hpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+ private:
+  const tuples::TaggedTupleTypelist<tag_list> const_global_cache_;
```
Comment (kidder) [inline src/Parallel/ConstGlobalCache.hpp]: remove const

File: tests/Unit/Parallel/Test_ConstGlobalCache.cpp [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+}  // namespace
+
+SPECTRE_TEST_CASE("Unit.Parallel.ConstGlobalCache", "[Unit][Parallel]") {
+  tuples::TaggedTupleTypelist<
+      typename TestMetavariables::const_global_cache_tag_list>
+      const_data_to_be_cached("Nobody", 178, 2.2, std::make_unique<Square>());
+
+  Parallel::ConstGlobalCache<TestMetavariables> cache(const_data_to_be_cached);
```
Comment (kidder) [inline tests/Unit/Parallel/Test_ConstGlobalCache.cpp:105]: std::move

File: tests/Unit/Parallel/Test_ConstGlobalCache.cpp [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    CHECK(-1 == retrieved_proxy.ckLocal()->my_id());
+  }
+  for (int i = 0; i < 40; ++i) {
+    if (nullptr != retrieved_array_proxy[i].ckLocal()) {
+      CHECK(i == retrieved_array_proxy[i].ckLocal()->my_index());
+    }
+  }
+  CHECK("Nobody" == cache.get<name>());
+  CHECK(178 == cache.get<age>());
+  CHECK(2.2 == approx(cache.get<height>()));
```
Comment (kidder) [inline tests/Unit/Parallel/Test_ConstGlobalCache.cpp]: no approx

File: tests/Unit/Parallel/Test_ConstGlobalCache.hpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  TestGroupChare() = default;
+  explicit TestGroupChare(CkMigrateMessage* /*unused*/) {}
+  int my_proc() const { return Parallel::my_proc(); }
+};
+
+class TestNodeGroupChare : public CBase_TestNodeGroupChare {
+ public:
+  TestNodeGroupChare() = default;
+  explicit TestNodeGroupChare(CkMigrateMessage* /*unused*/) {}
+  int my_node() const { return Parallel::my_node(); }
```
Comment (kidder) [inline tests/Unit/Parallel/Test_ConstGlobalCache.hpp]: noexcept stuff

File: tests/Unit/Parallel/Test_ConstGlobalCache.hpp [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  explicit TestArrayChare(CkMigrateMessage* /*unused*/) {}
+  int my_index() { return thisIndex; }
+};
+
+class TestChare : public CBase_TestChare {
+ public:
+  explicit TestChare(int id) : id_(id) {}
+  explicit TestChare(CkMigrateMessage* /*unused*/) {}
+  int my_id() const { return id_; }
+  int id_{std::numeric_limits<int>::max()};
```
Comment (kidder) [inline tests/Unit/Parallel/Test_ConstGlobalCache.hpp]: private

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**